### PR TITLE
Retries added to the schema registry

### PR DESCRIPTION
**Pull request by ZsoltBalvanyos**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -34,6 +34,7 @@ lazy val serialisation = Project("comms-kafka-serialisation", file("modules/seri
       "com.ovoenergy"       %% "kafka-serialization-core"   % "0.1.18",
       "com.ovoenergy"       %% "kafka-serialization-avro4s" % "0.1.18",
       "org.slf4j"           % "slf4j-api"                   % "1.7.21",
+      "com.typesafe.akka"   %% "akka-stream-kafka"          % "0.16",
```
Comment (cb372) [inline build.sbt]: why is this new dependency needed?

File: modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Config.scala [hunk 2]
```diff
@@ -1,7 +1,5 @@
-package com.ovoenergy.comms.helpers
+package com.ovoenergy.comms.serialisation
```
Comment (cb372) [inline modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Config.scala]: I would leave this file in the helpers module, and only move the `Retry` object to the serialisation module

File: modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Serialisation.scala [hunk 3]
```diff
@@ -191,25 +193,45 @@ object Serialisation {
...
-    schemaRegistryClient.register(s"$topic-value", schema)
+  private def registerSchema[T](schemaRegistryClient: SchemaRegistryClient,
+                                topic: String,
+                                schemaRegistryConfig: SchemaRegistryConfig)(implicit sf: SchemaFor[T]) = {
+
+    val schema                = sf.apply()
+    val trySchemaRegistration = () => Try(schemaRegistryClient.register(s"$topic-value", schema))
+
+    val onFailure = { (e: Throwable) =>
+      {
```
Comment (cb372) [inline modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Serialisation.scala]: you don't need the inner set of curly braces

File: modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Serialisation.scala [hunk 4]
```diff
@@ -191,25 +193,45 @@ object Serialisation {
...
+        log.debug(s"Schema registration attempt was unsuccessful. ${e.getMessage}")
+      }
+    }
+
+    Retry.retry(schemaRegistryConfig.retry, onFailure)(trySchemaRegistration) match {
+      case Left(l) => {
+        log.error("Schema registration failed! Killing the JVM.")
+        System.exit(1)
```
Comment (cb372) [inline modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Serialisation.scala]: I think killing the JVM here is a bit too brutal. We should probably make methods such as `avroBinarySchemaRegistrySerializer ` return an `Either[some kind of error info, Serializer[T]]` and let the caller decide what to do. Having a method in a library suddenly terminate the JVM is a bit surprising.

File: build.sbt (hunk 1 above)
Comment (ZsoltBalvanyos) [reply build.sbt]: It is needed for the Retry object.

File: modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Config.scala (hunk 2 above)
Comment (ZsoltBalvanyos) [reply modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Config.scala]: If I leave the Config file in the serialisation module, i get circular dependency problem between helpers and serialisation.

File: modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Serialisation.scala (hunk 4 above)
Comment (ZsoltBalvanyos) [reply modules/serialisation/src/main/scala/com/ovoenergy/comms/serialisation/Serialisation.scala]: ok

**Pull request merged**